Add new `Literal::str_value`, `Literal::cstr_value` and `Literal::byte_str_value` methods

GuillaumeGomez opened the pull request:
Since https://github.com/rust-lang/rust/pull/139367 was merged a while ago, I think it's ok to start making use of it as a nightly experiment if you're ok with it.

If this PR is merged, I plan to send a follow-up on `syn` for `Literal::value` to make use of these methods instead (when `nightly` feature is enabled) so the unescaping source code could be `cfg`-ed out.

Hopefully, when the API gets stabilized, we could get some nice numbers on compilation improvements for a lot of crates (although minor, considering how much your crates are used, definitely worth it).

So what do you think of this?

dtolnay reviewed (changes requested):
Thanks for the PR. Since the API of these methods is likely to change, in order to move this forward they need to be behind the `procmacro2_semver_exempt` cfg like the unstable `Span` methods such as `def_site` &mdash; a Cargo feature is not appropriate for this. Separately, each method needs a fallback implementation with identical behavior that works in non-macro code.

GuillaumeGomez commented:
Makes sense! A bit busy currently but I'll try to come back to this as soon as possible.

GuillaumeGomez commented:
Removed the feature and switched to the `cfg`, however I don't know how to make the `rustc-literal-escaper` dependency optional with a cfg. Or maybe you have another idea or how to unescape (like copy-pasting code from `syn`)?

dtolnay commented on Cargo.toml:
```diff
@@ -31,6 +31,9 @@ features = ["span-locations"]
 unicode-ident = "1.0"
 
+[target.'cfg(procmacro2_semver_exempt)'.dependencies]
+rustc-literal-escaper = "0.0.5"
```
This dependency's compiler version support is not compatible with making it a dependency of proc-macro2. Also it has unreasonably many publishers for something that would become a widely used dependency. Neither of these factors is necessarily immediately disqualifying while the dependency is gated by our semver exempt cfg, but they mean that this implementation is not going to work once it comes time to stabilize the new methods. I would prefer going directly to a different implementation that does not involve adding this dependency.

dtolnay commented on src/lib.rs:
```diff
@@ -1271,6 +1280,113 @@ impl Literal {
+            imp::Literal::Compiler(ref compiler_lit) => compiler_lit.str_value(),
+            imp::Literal::Fallback(ref fallback) => {
+                if !fallback.repr.starts_with('"') {
+                    return Err(ConversionErrorKind::InvalidLiteralKind);
```
Is this the intended behavior on raw string literals?

dtolnay commented on src/lib.rs:
```diff
@@ -88,6 +88,10 @@
 #![cfg_attr(docsrs, feature(doc_cfg))]
+#![cfg_attr(
+    all(procmacro2_semver_exempt, feature = "proc-macro"),
+    feature(proc_macro_value)
```
It is not worth introducing the use of an unstable feature for this API. That makes sense for things like Span::def_site and Literal::subspan where a stable implementation is impossible. But for the methods in this PR they should just follow the stable "fallback" implementation in both cases, without being coupled to the API in nightly.
